docs/proposals/L7-protocols.md

```diff
@@ -0,0 +1,190 @@
+  apiVersion: v1
+  kind: Service
+  metadata:
+    name: backend
+    namespace: example
+  spec:
+    ports:
+    - port: 8080
+      name: http          # `http` protocol in the `name`
+    - port: 7070
+      name: http-metrics  # `http` protocol in the `name`
+    - port: 6060
+      name: somethinggrpc # `grpc` protocol in the `name`
+    selector:
+      app: backend
+  ```
+
+* A user might opt for explicit annotations in the form `<port>.service.kuma.io/protocol`
+
+  ```yaml
+  apiVersion: v1
+  kind: Service
+  metadata:
+    name: backend
+    namespace: example
+    annotations:
+      8080.service.kuma.io/protocol: http # explicit `http` protocol
+      7070.service.kuma.io/protocol: http # explicit `http` protocol
+      6060.service.kuma.io/protocol: grpc # explicit `grpc` protocol
+  spec:
+    ports:
+    - port: 8080
+      name: a
+    - port: 7070
+      name: b
+    - port: 6060
+      name: tcp  # `protocol` from annotation will take precedence
+    selector:
+      app: backend
+  ```
+
+### Configuration model #2
+
+* User describes a high-level service, `Kuma` infers low-level technical aspects, such as `protocol`
+
+#### Universal mode
+
+* REST API
+  ```yaml
+  type: ServiceProfile
+  mesh: default
+  name: backend # the same value as in `service` tag
+  profiles:
+  - api:
+      rest:
+        spec:
+          url: https://generator.swagger.io/api/swagger.json
+  ```
+* gRPC service
+  ```yaml
+  type: ServiceProfile
+  mesh: default
+  name: backend # the same value as in `service` tag
+  profiles:
+  - api:
+      grpc:
+        spec:
+          url: https://github.com/envoyproxy/envoy/blob/master/api/envoy/rvice/accesslog/v3/als.proto
+  ```
+* Postgres database
+  ```yaml
+  type: ServiceProfile
+  mesh: default
+  name: db      # the same value as in `service` tag
+  profiles:
+  - database:
+      postgres: {}
+  ```
+
+#### Kubernetes mode
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: ServiceProfile
+metadata:
+  namespace: example
+  name: backend # the same name as `Service` resource has
+mesh: default
+spec:
+  profiles:
+  - api:
+      rest:
+        spec:
+          url: https://generator.swagger.io/api/swagger.json
+```
+
+## Other Considerations
+
+### About other ways to infer application protocol automatically
+
+* port numbers (e.g., `80`, `443`, `8080`, `5432`) and `k8s` health checks (`HTTP` and `TCP`) are another good strategy to infer application protocol 
+* however, they are not a part of this proposal
+* we could add support for it in the future, probably in the form of recommendations
+
+### About "protocol" tag in case of ServiceProfile
+
+* If it is important to have `protocol` tag for the sake of `Kuma` policy matching (e.g., to be able to apply a policy to all `http` services), this tag could be generated automatically out of a `ServiceProfile`
+
+## Open questions
+
+1. What happens if a user defines different `protocol` tag for the same `service` in different `Dataplanes` ?
```

Comment:
We need this information when generating inbound listeners, right? We generate listener for a given dataplane, so if one DP is defined with TCP then TCP listener will be generated, if DP is defined with HTTP then HTTP listener will be generated.